File: leetcode_411.cpp
class Solution {
public:
  bool isValid (string const& s) {
    string left = "([{";
    string right = ")]}";
    stack<char> stk;

    for (auto c : s) {
      if (left.find(c) != string::npos) {
        stk.push(c);
      } else {
        if (stk.empty() || stk.top != left[right.find(c)]) {
          return false;
        } else {
          stk.pop();
        }
      }
    }
    return stk.empty();
  }
}

File: leetcode_411.py
class Solution:
    # @return a boolean
    def isValid(self, s):
        stack = []
	for c in s:
	  if len(stack) > 0:
       	    tmp = stack[len(stack) - 1]
	  else:
	    tmp = None
	  if c == '(' or c == '[' or c == '{':
	    stack.append(c)
	  elif c == ')':
	    if tmp != '(':
	      return False 
	    else:
	      stack.pop()
	  elif c == ']':
	    if tmp != '[':
	      return False
	    else:
	      stack.pop()
	  elif c == '}':
	    if tmp != '{':
	      return False
	    else:
	      stack.pop()

	return len(stack) == 0

s = Solution()
print s.isValid('()')
print s.isValid('()[]{}')
print s.isValid('(]')
print s.isValid('([)]')
